File: Practice/Implementation/q14/q14.py
from itertools import permutations

def solution(n, weak, dist):
    length = len(weak)
    for i in range(length):
        weak.append(weak[i]+n)
    print(list(permutations(dist, len(dist))))
    answer = len(dist)+1
    for start in range(length):
        for friends in list(permutations(dist, len(dist))):
            count = 1
            position = weak[start] + friends[count-1]
            for index in range(start, start+length):
                if position < weak[index]:
                    count += 1
                    if count > len(dist):
                        break
                    position = weak[index] + friends[count-1]
            answer = min(answer, count)
    return -1 if answer > len(dist) else answer

File: Practice/Implementation/q14/q14.cpp
#include <string>
#include <algorithm>
#include <vector>

using namespace std;

int solution(int n, vector<int> weak, vector<int> dist) {
    int length = weak.size();
    for(int i = 0; i < length; i++){
        weak.push_back(weak[i] + n);
    }
    int answer = dist.size() + 1;
    for(int start = 0; start < length; start++){
        do{
            int cnt = 1;
            int position = weak[start] + dist[cnt-1];
            for(int index = start; index < start+length; index++){
                if(position < weak[index]){
                    cnt++;
                    if(cnt > dist.size()){
                        break;
                    }
                    position = weak[index] + dist[cnt-1];
                }
            }
            answer = min(answer, cnt);
        }while(next_permutation(dist.begin(), dist.end()));
    }
    if(answer > dist.size()){
        return -1;
    }
    return answer;
}